### ISSUE TITLE
Handshake failure due unexpected packet format

### ISSUE THREAD

**Issue by Mak-ING** [JeffFerguson/gepsio#20]
Hello,
We are using JeffFerguson.Gepsio library since quit a long time. Recently we are getting an error when we try to load any XBRL document. The error comes when it tries to read some xsd references mentioned in the document. 
Example  "https://eba.europa.eu/eu/fr/xbrl/crr/fws/finrep/cir-680-2014/2017-04-04/mod/finrep9_con_ifrs.xsd"
It will be highly appreciated if you could provide some insights on this issue.
The deailed error log is as below.
2019-01-09 10:28:41,467 [4] [User] ERROR - System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a send. ---> System.IO.IOException: The handshake failed due to an unexpected packet format.
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.TlsStream.CallProcessAuthentication(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Net.TlsStream.ProcessAuthentication(LazyAsyncResult result)
   at System.Net.TlsStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.PooledStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.ConnectStream.WriteHeaders(Boolean async)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.GetResponse()
   at System.Xml.XmlDownloadManager.GetNonFileStream(Uri uri, ICredentials credentials, IWebProxy proxy, RequestCachePolicy cachePolicy)
   at System.Xml.XmlDownloadManager.GetStream(Uri uri, ICredentials credentials, IWebProxy proxy, RequestCachePolicy cachePolicy)
   at System.Xml.XmlUrlResolver.GetEntity(Uri absoluteUri, String role, Type ofObjectToReturn)
   at System.Xml.XmlTextReaderImpl.FinishInitUriString()
   at System.Xml.XmlTextReaderImpl..ctor(String uriStr, XmlReaderSettings settings, XmlParserContext context, XmlResolver uriResolver)
   at System.Xml.XmlReaderSettings.CreateReader(String inputUri, XmlParserContext inputContext)
   at System.Xml.XmlReader.Create(String inputUri, XmlReaderSettings settings, XmlParserContext inputContext)
   at System.Xml.XmlReader.Create(String inputUri)
   at JeffFerguson.Gepsio.Xml.Implementation.SystemXml.Schema.Read(String path) in c:\Users\User\Path\Sources\BusinessLayer\DomainObjects.XBRL\Xml\Implementation\SystemXml\Schema.cs:line 63
   at JeffFerguson.Gepsio.XbrlSchema..ctor(XbrlFragment ContainingXbrlFragment, String SchemaFilename, String BaseDirectory) in c:\Users\user\Projects\NET\Sources\BusinessLayer\DomainObjects.XBRL\XbrlSchema.cs:line 121

where path = "https://eba.europa.eu/eu/fr/xbrl/crr/fws/finrep/cir-680-2014/2017-04-04/mod/finrep9_con_ifrs.xsd"

**Comment by JeffFerguson** [comment]
Thank you for the report! Does the XBRL instance that you're loading reference this schema directly, or is this schema loaded on behalf of another reference? Would it be possible for you to send over the XBRL instance that you're loading when this error appears?

**Comment by Mak-ING** [comment]
Hello,
Thank you for the quick response. I am attaching the XBRL instance we are trying to load. I have changed its extension from .xbrl to .txt because files with xbrl extension cannot be uploaded.

[DUMMYLEI123456789012_GB_FINREP020202_FINREP9CONIFRS_2018-12-31_20180802120230000.txt](https://github.com/JeffFerguson/gepsio/files/2741464/DUMMYLEI123456789012_GB_FINREP020202_FINREP9CONIFRS_2018-12-31_20180802120230000.txt)

**Comment by JeffFerguson** [comment]
Thank you! I'll look into this issue for you as soon as possible.

**Comment by Mak-ING** [comment]
Hi Jeff,
Any update on this issue. Our internal investigation shows as follows:
“System.Net.WebException: The request was aborted: Could not create SSL/TLS secure channel.”

I have checked the remote webserver and noticed it has a very strict SSL/TLS configuration, please re-run the full analysis via: 
https://www.ssllabs.com/ssltest/analyze.html?d=eba.europa.eu 

So it's most likely that the current .NET application code and/or libraries are not compatible with the required TLS v1.2 which is enforced by the remote webserver in order to communicate.
 
Long story short, kindly can you please comment on the TLS support by your library. Exact place in the code where we can check and change this? 
Your help will be highly appreciated.
Thank You

**Comment by JeffFerguson** [comment]
Good morning. Sorry for the delay. Gepsio does not open sockets directly, and relies on the code in the .NET Framework's `System.Xml` namespace for all HTTP-based data retrieval. I need to do a bit more research in finding out how that code supports TLS 1.2. I will have some time today to look into this for you.

**Comment by Mak-ING** [comment]
Hello,
Thanks for you quick reply. The issue is resolved now. It was indeed related to socket connections. I upgraded all my projects from .net framework 4.5 to 4.6.1 (which uses TLS v1.2 by default for web requests). Working fine. 
Thank you for the support.
